### PR TITLE
DB-9770/DB-10267 Fix External Table Row Counts after cross-join insert

### PR DESCRIPTION
**Pull request by martinrupp**
consisting of changes DB-9770 (fix) and DB-10267 (speed improvement)

### CODE REVIEW

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java [hunk 1]
```diff
@@ -1114,64 +1116,98 @@ public static boolean joinProjectsOnlyLeftTableColumns(JoinType joinType) {
...
+        return getRowsWritten(context);
     }
 
-    private DataFrameWriter getDataFrameWriter(int[] partitionBy, String compression, OperationContext context) throws StandardException {
+    class CountingListener extends SparkListener implements AutoCloseable
+    {
+        OperationContext context;
+        SparkContext sc;
+        String uuid;
+        List<Integer> stageIdsToWatch;
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java:1176]: Can you make it a Set? Since we are going to call lots of `.contains()` on it I think that makes sense.

File: hbase_sql/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/ExternalTableIT.java [hunk 2]
```diff
@@ -2409,15 +2446,15 @@ public void testParquetPartitionColumnName() throws Exception {
     public void testOrcColumnName() throws Exception {
         String tablePath = getExternalResourceDirectory()+"orc_colname";
         methodWatcher.execute(String.format("create external table t_orc (col1 int, col2 varchar(5))" +
-                " STORED AS ORC LOCATION '%s'", tablePath));
+                " STORED AS PARQUET LOCATION '%s'", tablePath));
```
Comment (dgomezferro) [inline hbase_sql/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/ExternalTableIT.java:2449]: Why the change? If this is deliberate we'd need to change the test name too

**Review by dgomezferro — CHANGES_REQUESTED** [review]

File: hbase_sql/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/ExternalTableIT.java (hunk 2 above)
Comment (martinrupp) [reply hbase_sql/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/ExternalTableIT.java:2449]: good catch, that wasn't intentional.

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java (hunk 1 above)
Comment (martinrupp) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java:1176]: yes will do

**Review by arnaud-splice — APPROVED** [review]
No additional comments once Daniel's are fixed.
Good PR!

**Review by dgomezferro — APPROVED** [review]
Looks good, thanks!

**Comment by martinrupp** [comment]
this needs a fix to 10232 to work correctly

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1610/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1614/

**Review by carolp-503 — APPROVED** [review]
+1
